message/router.go

```diff
@@ -158,25 +185,54 @@ func (r *Router) AddHandler(
 		return errors.Errorf("handler %s already exists", handlerName)
 	}
 
+	publisherName, subscriberName := r.pubSubNames(publisher, subscriber)
+
 	r.handlers[handlerName] = &handler{
 		name:   handlerName,
 		logger: r.logger,
 
 		subscriber:     subscriber,
 		subscribeTopic: subscribeTopic,
-		publisher:      publisher,
-		publishTopic:   publishTopic,
+		subscriberName: subscriberName,
+
+		publisher:     publisher,
+		publishTopic:  publishTopic,
+		publisherName: publisherName,
 
-		runningHandlersWg: r.runningHandlersWg,
 		handlerFunc:       handlerFunc,
+		runningHandlersWg: r.runningHandlersWg,
 		messagesCh:        nil,
-
-		closeCh: r.closeCh,
+		closeCh:           r.closeCh,
 	}
 
 	return nil
 }
 
+// pubSubNames resolves the names of publisher and subscriber.
+// They are then stored in the context of the message.
+func (r *Router) pubSubNames(pub Publisher, sub Subscriber) (string, string) {
```

Comment:
why not two functions? :P

Reply:
because a0fbd1782a8e7f2273b4130000356a3c3ab03ad3